### PR TITLE
Let `document.hasStorageAccess` check whether the Document already has unpartitioned data access

### PR DESCRIPTION
**Pull request by shuranhuang**
This commits tries to make hSA match the description in the spec that “This specification defines a method to query whether or not a [Document](https://dom.spec.whatwg.org/#document) currently has access to its [unpartitioned data](https://privacycg.github.io/storage-access/#unpartitioned-data) ([hasStorageAccess()](https://privacycg.github.io/storage-access/#dom-document-hasstorageaccess)) …” by including a check of whether the user agent allows the `document` to access unpartitioned data based on user settings.

- [x] At least two implementers are interested (and none opposed)
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://chromium-review.googlesource.com/c/chromium/src/+/4611289
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://bugs.chromium.org/p/chromium/issues/detail?id=1433013
   * Gecko: N/A
   * WebKit: N/A

Fixes https://github.com/privacycg/storage-access/issues/171

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

***
<a href="https://pr-preview.s3.amazonaws.com/shuranhuang/storage-access/pull/174.html" title="Last updated on Jul 25, 2023, 2:30 PM UTC (024f4b8)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/privacycg/storage-access/174/69042e6...shuranhuang:024f4b8.html" title="Last updated on Jul 25, 2023, 2:30 PM UTC (024f4b8)">Diff</a>

### CODE REVIEW

File: storage-access.bs
Comment (annevk) [inline storage-access.bs:450]: Restore that newline please.

File: storage-access.bs
Comment (annevk) [inline storage-access.bs:193]: Resolve has a required argument, this change is incorrect.

File: storage-access.bs
Comment (annevk) [inline storage-access.bs:132]: We don't want to require these settings to be global. I'd suggest removing this.

File: storage-access.bs
Comment (annevk) [inline storage-access.bs:167]: We should continue to queue a task for this case. But also, this seems wrong. You can have storage access but no cookie access, right? E.g., for top-level documents that seems possible.

File: storage-access.bs
Comment (annevk) [inline storage-access.bs:169]: I think we should look at whether the user agent is configured to allow access to cookies for the (top-level site, embedded site) combination. Or something like that, perhaps it needs to be implementation-defined.

**Review by annevk — COMMENTED** [review]
Thanks for tackling this, but I think this needs a bit more work.

File: storage-access.bs
Comment (shuranhuang) [reply storage-access.bs:167]: > You can have storage access but no cookie access, right? E.g., for top-level documents that seems possible.

Thanks for the review, Anne! To make sure I understand correctly, could you help clarify what you mean with this comment? Do you mean 
- "the global's `has storage access` bit could be true but the user could have changed settings so that cookies are not really accessible", or 
- "the document could have dom storage access but not cookie access "?

Comment (annevk) [reply storage-access.bs:167]: The first of those. I think that could also happen for a cross-site document, if you had previously given storage access, but changed settings since then to block cookies.

**Review by annevk — COMMENTED** [review]
This looks better, but:

1. The state you grab from the global you could grab directly in the steps of the task. With what you do now you look at a cached value, which I guess theoretically could have changed. If you meant to do that, a note would help.
2. The user agent settings shouldn't be grabbed as part of the task steps. You want to grab them while in parallel.
3. Is "unpartitioned data" correct? We're just dealing with cookies here, right?

File: storage-access.bs
Comment (annevk) [inline storage-access.bs:173]: You need to grab the state from the global inside the task steps, not inside the in parallel steps.

And instead of sending all the user agent settings over, why not check for the particular bit you care about while in parallel and send the result of that over?

File: storage-access.bs
Comment (annevk) [inline storage-access.bs:166]: Instead of `<dfn>` you want `<var>` with camel casing as this is an inline definition.

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:169]: As a drive-by note, I think it would be nice to use something like `if the user agent's settings explicitly allow unpartitioned cookie access for |doc|` (emphasis on the *explicit* here) and then have a non-normative note informing where such an explicit allow could come from (i.e. the user changing global browser settings or per-site allow-lists).

Comment (annevk) [reply storage-access.bs:169]: Should it be _doc_ or some kind of (site, site) tuple? With the former it doesn't seem like the kind of thing you'd want to do in parallel.

Comment (johannhof) [reply storage-access.bs:169]: I think I asked you this before but are there rules around which concepts we can reference in parallel vs. not? It feels like |doc| may be enough information to allow the implementer to infer that it should be whatever reference to the doc the main process holds, but OTOH it's also easy to extract the (site, site) pair from the doc ahead of time so this is more of a principle question :)

Comment (annevk) [reply storage-access.bs:169]: No exact rules. But making things somewhat explicit is generally preferred.

Comment (shuranhuang) [reply storage-access.bs:169]: Thanks for the input! Is there any existing `(site, site) tuple` use case I can take a look to get some idea of the writing format?

Comment (annevk) [reply storage-access.bs:169]: It's what we currently use as permission key type: https://privacycg.github.io/storage-access/#permissions-integration.

Comment (shuranhuang) [reply storage-access.bs]: Still looking into how to rephrase with mentioning `(site, site) tuple` instead of `doc`. But could you PTAL the current steps first? Thanks!

**Comment by Rickyoperio** [comment]
Yes?

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:168]: nit: is it necessary to capitalize every word of the algorithm? It reads a bit unnaturally to me :)

Comment (johannhof) [inline storage-access.bs:168]:
```suggestion
    1. Let |Whether the User Agent Explicitly Allows Unpartitioned Cookie Access| be an algorithm that, given a [=tuple=] |tuple| consisting of [=sites=], runs the following steps. This algorithm returns "none", "allow" or "disallow".
```

It also seems a bit unconventional to include the "this algorithm returns" but I don't think it hurts?

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:170]: I think I'd phrase it like this:

```suggestion
        Note: A user agent's settings might explicitly allow or disallow unpartitioned cookie access through per-site allow-lists, the user changing global browser settings, or similar custom overrides.
```

I also wonder if we need to explicitly mention that this is implementation-defined, but OTOH this note is non-normative anyway, and the text also implies it.

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:175]:
```suggestion
    1. Let |explicitSetting| be the result of determining |Whether the User Agent Explicitly Allows Unpartitioned Cookie Access| with (|topLevelSite|, |embeddedSite|).
```

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:176]: Hmm I'm still a bit unsure why we need the permission check here, sorry 😳 , shouldn't this be determined by the has storage access bit? Happy to chat about this offline and follow up here.

File: storage-access.bs
Comment (shuranhuang) [reply storage-access.bs]: "This algorithm returns" is also used in [permissions spec](https://w3c.github.io/permissions/), so I thought it is fine to use here.

File: storage-access.bs
Comment (johannhof) [reply storage-access.bs:176]: Clarified with Shuran that this is for controlling access where the access may have been revoked by the user but the iframe still has storage access. In that case it looks like we want the user decision to immediately override the cookie access without pending a reload on the iframe or similar.

It might be good to add a non-normative note to that effect?

I should note that things won't work that way if we add storage support in #102 because that will probably exist for the lifetime of the document, unless we add in some mechanism to sever access once permission is revoked (I'm not sure that's a good idea though).

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:172]:
```suggestion
        1. If the user agent does not have explicit settings for unpartitioned cookie access for |tuple|, return "none".
```

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:173]:
```suggestion
        1. If the user agent's settings explicitly allow unpartitioned cookie access for |tuple|, return "allow".
```

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:174]:
```suggestion
        1. If the user agent's settings explicitly disallow unpartitioned cookie access for |tuple|, return "disallow".
```

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:176]: Do you need to call it "previous" permission state? Is it being replaced by anything? Otherwise `permissionState` or `currentPermissionState` should work.

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:181]:
```suggestion
                Note: The global storage access permission state takes precedence over the local [=environment/has storage access=] flag here, in order to immediately reflect a possible user choice to revoke the permission in their settings.
```

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:179]: No need to return |p| in the "queue a global task" steps.

Comment (shuranhuang) [reply storage-access.bs]: I see, thanks!

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:179]: Could this be simplified to `If |explicitSetting| is "none" and |permissionState| is ...`?

File: storage-access.bs
Comment (johannhof) [inline storage-access.bs:174]: I guess "disallow" is unused here now, but we might use it if we re-use this algorithm for rSA later?

**Review by johannhof — APPROVED** [review]
LGTM but would like to wait on @annevk's review before merging.

File: storage-access.bs
Comment (shuranhuang) [reply storage-access.bs:174]: That's right. "disallow" is not checked here explicitly, but we can add

> 1. If |explicitSetting| is "disallow", [=/resolve=] |p| with false.

if that's clearer/preferred.

Comment (johannhof) [reply storage-access.bs:174]: Hmm not sure, it seems unnecessary to add there. It's probably fine like this? Maybe @annevk has a preference...

**Comment by shuranhuang** [comment]
Friendly ping on this @annevk :)

File: storage-access.bs
Comment (annevk) [reply storage-access.bs:174]: I'm somewhat surprised this has explicit "allow". I thought that was always subject to `requestStorageAccess()` for child frames?

**Review by annevk — COMMENTED** [review]
Didn't we want to make `hasStorageAccess()` reflect deny states for the top-level document as well? It seems earlier steps of `hasStorageAccess()` end up returning early with true.

**Comment by johannhof** [comment]
> Didn't we want to make `hasStorageAccess()` reflect deny states for the top-level document as well? It seems earlier steps of `hasStorageAccess()` end up returning early with true.

Hmm I don't think that was an expectation on my end. I do see SAA primarily as a mechanism governing cross-site data access, not same-site (or same authority). FWIW I'd be okay with punting on a follow-up.

File: storage-access.bs
Comment (johannhof) [reply storage-access.bs:174]: There are user agent settings that disable 3PC blocking (e.g. in Safari IIRC) or allow-list specific sites (e.g. in Firefox and Chrome), how else would we represent those?

**Comment by shuranhuang** [comment]
@annevk does it work for you if I create a follow-up PR for the top-level document / same authority case? I can update the PR title to "Let document.hasStorageAccess check whether the **embedded** Document already has unpartitioned data access"

**Comment by shuranhuang** [comment]
@annevk PTAL the latest version.

File: storage-access.bs
Comment (annevk) [inline storage-access.bs:164]: consisting of two sites*

File: storage-access.bs
Comment (johannhof) [reply storage-access.bs:177]: Yes, this still LGTM :)

File: storage-access.bs
Comment (shuranhuang) [reply storage-access.bs:172]: That's right. `permissionState` is only checked when `explicitSetting` is "`none`".

Comment (shuranhuang) [reply storage-access.bs]: That's right, done.

**Comment by shuranhuang** [comment]
@annevk does it look good with your suggestions applied?

**Comment by shuranhuang** [comment]
@annevk  Friendly ping:)

**Comment by johannhof** [comment]
In the interest of moving things along on our graduation goals, I'll go ahead and merge this without @annevk's explicit sign-off, as there's been plenty of review (and a final positive note). I hope that works for you, Anne. I think @shuranhuang is happy to correct any remaining concerns you may have in another PR.

**Pull request merged**
